### PR TITLE
aruha-856 authorize stream read

### PR DESCRIPTION
**Pull request by rcillo**
Every endpoint related to reading from a partition must be
authorized. Not only getting events, but getting information about how
many events there are or any sort of glimpse over a partition.

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.zalando.nakadi.plugin.api.authz.AuthorizationService;
+import org.zalando.nakadi.plugin.api.authz.Resource;
+import org.zalando.nakadi.plugin.api.authz.Subject;
+import org.zalando.nakadi.repository.EventTypeRepository;
+import org.zalando.nakadi.security.Client;
+
+import java.util.List;
+import java.util.Optional;
+
+public class AuthorizationUtils {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: We have AuthorizationValidator in master. It is implemented as bean, I think it is better to use it.

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java [hunk 2]
```diff
@@ -154,11 +156,15 @@ public StreamingResponseBody streamEvents(
...
                         nakadiSettings.getDefaultCommitTimeoutSeconds(), client.getClientId());
                 streamer = subscriptionStreamerFactory.build(subscriptionId, streamParameters, output,
-                        connectionReady, blacklistService);
+                        connectionReady, blacklistService, client);
                 streamer.stream();
             } catch (final InterruptedException ex) {
                 LOG.warn("Interrupted while streaming with " + streamer, ex);
                 Thread.currentThread().interrupt();
+            } catch (final AccessDeniedException e) {
+                writeProblemResponse(response, output.getOutputStream(), Problem.valueOf(Response.Status.FORBIDDEN,
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: maybe it is better to put this logic in SubscriptionOutputImpl.onException it will be simpler to refactor exceptions later

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                                                 final AuthorizationService authorizationService,
+                                                 final Client client, final SubscriptionBase subscriptionBase) {
+        subscriptionBase.getEventTypes().stream().forEach(
+                (eventTypeName) -> {
+                    try {
+                        eventTypeRepository.findByNameO(eventTypeName).ifPresent(eventType -> {
+                            authorizeStreamRead(authorizationService, client, eventType);
+                        });
+                    } catch (final InternalNakadiException e) {
+                        e.printStackTrace();
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: If we failed to check permissions, we should fail streaming, otherwise it is a security threat.

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            }
+
+            @Override
+            public String getType() {
+                return "event-type";
+            }
+
+            @Override
+            public Optional<List<AuthorizationAttribute>> getAttributesForOperation(
+                    final AuthorizationService.Operation operation) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: Interface is supposed to work with operation, but you do not use it :)

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java [hunk 5]
```diff
@@ -56,17 +65,23 @@ public SubscriptionStreamerFactory(
...
             final SubscriptionOutput output,
             final AtomicBoolean connectionReady,
-            final BlacklistService blacklistService) throws NoSuchSubscriptionException, ServiceUnavailableException,
+            final BlacklistService blacklistService,
+            final Client client) throws NoSuchSubscriptionException, ServiceUnavailableException,
             InternalNakadiException, NoSuchEventTypeException {
 
         final Subscription subscription = subscriptionDbRepository.getSubscription(subscriptionId);
+
+        authorizeSubscriptionRead(eventTypeRepository, authorizationService, client, subscription);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java]: The permission check is inside build method, which is to build the object, not to check the permission. It would be nice to have it in service layer, but we do not have it exactly here. We can move it to controller level, but there we need subscription. Maybe we can move retrieval of subscription to controller level or create service level for that?

File: src/main/java/org/zalando/nakadi/controller/CursorOperationsController.java [hunk 6]
```diff
@@ -54,21 +56,29 @@
...
     @RequestMapping(path = "/event-types/{eventTypeName}/cursor-distances", method = RequestMethod.POST)
     public ResponseEntity<?> getDistance(@PathVariable("eventTypeName") final String eventTypeName,
                                          @Valid @RequestBody final ValidListWrapper<CursorDistance> queries,
-                                         final Client client) {
+                                         final Client client) throws InternalNakadiException, NoSuchEventTypeException {
+        // TODO: remove once new authorization is in place
         checkReadScopes(eventTypeName, client);
 
+        final EventType eventType = eventTypeRepository.findByName(eventTypeName);
+        authorizeStreamRead(authorizationService, client, eventType);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/CursorOperationsController.java]: shouldn't be these calls on the service layer?

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 3 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: god damn code generation. Forgot to rethrow c58a792 thanks for the catch

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java (hunk 5 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java]: I'll extract subscription retrieval to the controller.

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java [hunk 7]
```diff
@@ -55,7 +56,7 @@ public static void authorizeSubscriptionRead(final EventTypeRepository eventType
                             authorizeStreamRead(authorizationService, client, eventType);
                         });
                     } catch (final InternalNakadiException e) {
-                        e.printStackTrace();
+                        throw new MyNakadiRuntimeException1("Failed authorizing");
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: As we agreed not to use MyNakadiRuntimeException1 for throwing, I suggest ServiceTemporarilyUnavailableException

File: src/main/java/org/zalando/nakadi/controller/CursorOperationsController.java (hunk 6 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/CursorOperationsController.java]: I don't know. I gave it some thought and it seems wrong to have to edit every single method signature to include a client. It feels like coupling the execution of an operation with checking if that operation is authorized. It might be that I already have a fixed idea in mind because when I implemented this kind of feature before, it would usually be implemented using some aspects or something outside of the actual operation.

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java (hunk 2 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: I considered doing this but I would have to check the type of exception there. Do you think it would be fine?

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java (hunk 5 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java]: Done a82a804

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 4 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: yeah. I'll maybe just use the resource you created in your PR. I'll wait for it to be merged and then fix it here, what do you think?

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 7 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: 👍  d0c2979

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 1 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: I thought this class is responsible for validating some parameters and not actually implementing code that performs authorization.

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java (hunk 2 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: Adding TODO about it would be good

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 4 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: it is already in master, but I propose to use resource which was created by @antban, it is `almost` in master :)

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 1 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: One could think of it as of permissions validation :) anyway, it is better to inject object rather than using helpers, in the end we use spring

File: src/main/java/org/zalando/nakadi/controller/PartitionsController.java [hunk 8]
```diff
@@ -102,9 +117,13 @@ public PartitionsController(final TimelineService timelineService,
     public ResponseEntity<?> getPartition(@PathVariable("name") final String eventTypeName,
                                           @PathVariable("partition") final String partition,
                                           @Nullable @RequestParam(value = "consumed_offset", required = false)
-                                              final String consumedOffset, final NativeWebRequest request) {
+                                              final String consumedOffset, final NativeWebRequest request,
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/controller/PartitionsController.java]: broken formatting

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java [hunk 9]
```diff
@@ -59,16 +56,15 @@ public SubscriptionStreamerFactory(
...
             final AtomicBoolean connectionReady,
-            final BlacklistService blacklistService) throws NoSuchSubscriptionException, ServiceUnavailableException,
+            final BlacklistService blacklistService,
+            final Client client) throws NoSuchSubscriptionException, ServiceUnavailableException,
             InternalNakadiException, NoSuchEventTypeException {
-
-        final Subscription subscription = subscriptionDbRepository.getSubscription(subscriptionId);
         final Session session = Session.generate(1);
-        final String loggingPath = "subscription." + subscriptionId + "." + session.getId();
+        final String loggingPath = "subscription." + subscription.getId() + "." + session.getId();
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java]: what is the purpose of this change?

Comment (rcillo) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionStreamerFactory.java]: @v-stepanov to move the authorization outside of the builder, which is a very weird layer to make authorization.

File: src/main/java/org/zalando/nakadi/controller/PartitionsController.java (hunk 8 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/PartitionsController.java]: Not actually, this is a continuation of the line above, right?

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 4 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: Done 5c947b1

File: src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java (hunk 1 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/util/AuthorizationUtils.java]: Done

File: src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java (hunk 2 above)
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/controller/SubscriptionStreamController.java]: Done dfa5fe5

**Comment by rcillo** [comment]
@adyach I have addressed the comments

**Comment by adyach** [comment]
👍

**Comment by rcillo** [comment]
@adyach I accidentally committed a file which I just removed 95b1491

**Comment by adyach** [comment]
👍

**Comment by rcillo** [comment]
👍
